File: server/game.py
import time
from collections import deque

from easyAI import AI_Player, Negamax, TwoPlayersGame
from easyAI.AI import TT
from funcy import collecting

from board import display
from constants import (
    GOAT_CHAR,
    GOAT_PLAYER,
    NUM_EATEN_LOSE,
    NUM_GOATS,
    STARTING_TIGERS,
    TIGER_CHAR,
    TIGER_PLAYER,
)
from graph import JUMPS_GRAPH, STEPS_GRAPH
from ls import log
from notations import ALL_POS_NUMS, notation_to_pos_num, pos_num_to_notation
from pieces import Pieces


def apply_move(coords, pieces):
    if len(coords) == 1:
        pos = coords[0]
        pieces[pos] = GOAT_CHAR
    elif len(coords) == 2:
        src, dest = coords
        pieces[dest] = pieces.pop(src)
    elif len(coords) == 3:
        src, eaten, dest = coords
        pieces[dest] = pieces.pop(src)
        del pieces[eaten]


def unapply_move(coords, pieces):
    if len(coords) == 1:
        pos = coords[0]
        del pieces[pos]
    elif len(coords) == 2:
        src, dest = coords
        pieces[src] = pieces.pop(dest)
    elif len(coords) == 3:
        src, eaten, dest = coords
        pieces[src] = pieces.pop(dest)
        pieces[eaten] = GOAT_CHAR


class TigerAndGoat(TwoPlayersGame):
    history_max_len = 20

    def __init__(self, players):
        self.players = players
        self.pieces = self.initial_pieces()
        self.goats_to_place = NUM_GOATS
        # goat always starts
        self.nplayer = GOAT_PLAYER
        # check for repetition against the most recent states.
        # just far back enough to prevent endless loops.
        # AI lookahead will append and then pop (upon unmove), so this must be sized as
        # "enough history" + AI lookahead depth
        self.history = deque([self.pieces.canonical], maxlen=self.history_max_len)

    def ttentry(self, json_safe=False):
        """Game state consists of:
        - current player name
        - number of goats left to place
        - possibly truncated piece position history, always at least 1 turn back
            - while still placing goats just keep 1 turn, as move loops won't last
        """
        history = (self.history[-1],) if self.goats_to_place else tuple(self.history)
        if json_safe:
            history = tuple(list(map(list, pieces)) for pieces in history)
        return (self.nplayer, self.goats_to_place, history)

    def as_dict(self):
        nplayer, goats_to_place, history = self.ttentry()
        tigers, goats = history[-1]

        return {
            'playerNum': nplayer,
            'numGoatsToPlace': goats_to_place,
            'possibleMoves': self.possible_moves(),
            'history': history,
            'tigers': sorted(tigers),
            'goats': sorted(goats),
        }

    def ttrestore(self, entry):
        self.nplayer, self.goats_to_place, history = entry

        canonical = history[-1]

        self.pieces = Pieces(*canonical)
        self.history = deque(history, maxlen=self.history_max_len)

    def initial_pieces(self):
        # tigers in corner positions
        return Pieces(tigers=STARTING_TIGERS)

    def pretty_move(self, pos_nums):
        return pos_num_to_notation(pos_nums)

    def parse_move(self, move):
        try:
            return notation_to_pos_num(move)
        except ValueError:
            raise

    @log
    def make_move(self, coords):
        apply_move(coords, self.pieces)
        if len(coords) == 1:
            self.goats_to_place -= 1
        self.history.append(self.pieces.canonical)

    @log
    def unmake_move(self, coords):
        unapply_move(coords, self.pieces)
        if len(coords) == 1:
            self.goats_to_place += 1
        self.history.pop()

    def tigers_go(self):
        return self.nplayer == TIGER_PLAYER

    def goats_go(self):
        return self.nplayer == GOAT_PLAYER

    def empty(self):
        return ALL_POS_NUMS - set(self.pieces)

    def unique_positions(self):
        "Given a single goat's rotational and mirror symmetries"
        # For placing the first goat: use the upper half of the top-left quadrant
        return {0, 1, 2, 6, 7, 12}

    def is_repeat(self, coords):
        if self.goats_to_place:
            return False
        new_pieces = self.pieces.copy()
        apply_move(coords, new_pieces)
        return new_pieces.canonical in self.history

    def get_steps(self, nodes, short_circuit=False):
        for node in nodes:
            for dest in STEPS_GRAPH[node]:
                if dest not in self.pieces:
                    coords = node, dest
                    if not self.is_repeat(coords):
                        yield coords
                        if short_circuit:
                            break

    def tiger_steps(self, short_circuit):
        yield from self.get_steps(self.pieces.inverse[TIGER_CHAR], short_circuit)

    def tiger_jumps(self, short_circuit):
        for tiger in self.pieces.tigers:
            for dest in JUMPS_GRAPH[tiger]:
                eaten = (tiger + dest) // 2
                if self.pieces.get(eaten) == GOAT_CHAR and dest not in self.pieces:
                    coords = tiger, eaten, dest
                    yield coords
                    if short_circuit:
                        break

    def mobile_tigers(self):
        return len({mv[0] for mv in self.tiger_moves(short_circuit=True)})

    def goat_placements(self):
        # first goat placed only need consider unique positions
        if self.pieces == STARTING_TIGERS:
            positions = self.unique_positions() - self.pieces.inverse[TIGER_CHAR]
        else:
            positions = self.empty()
        return [(pos,) for pos in positions]

    def goat_steps(self):
        yield from self.get_steps(self.pieces.inverse[GOAT_CHAR])

    def tiger_moves(self, short_circuit=False):
        yield from self.tiger_jumps(short_circuit)
        yield from self.tiger_steps(short_circuit)

    def goat_moves(self):
        if self.goats_to_place:
            yield from self.goat_placements()
        else:
            yield from self.goat_steps()

    @log
    @collecting
    def possible_moves(self):
        if self.tigers_go():
            yield from self.tiger_moves()
        else:
            yield from self.goat_moves()

    def goats_eaten(self):
        return NUM_GOATS - self.goats_to_place - len(self.pieces) + 4

    def tiger_wins(self):
        return self.goats_eaten() >= NUM_EATEN_LOSE

    def goat_wins(self):
        return self.tigers_go() and not self.mobile_tigers()

    def is_draw(self):
        """If tiger can't move it loses. If goat can't move, we call it a draw,
        because we'd rather not incentivise this non-decisive behaviour
        """
        if self.goats_to_place or self.tigers_go():
            return False
        return not list(self.goat_moves())

    def show(self):
        display(self.pieces, self.goats_to_place, self.goats_eaten())
        if self.tiger_wins():
            print('5 goats eaten, TIGER WINS!')
        elif self.goat_wins():
            print('tiger cannot move, GOAT WINS!')
        elif self.is_draw():
            print('goat cannot move, A DRAW!')
        score = abs(self.scoring())
        print('score:', score, 'x' * score)

    @log
    def scoring(self):
        mobile_tigers = self.mobile_tigers()
        # if tiger just made a move, history may appear to preclude the reverse move,
        # however once goat has moved, this may not be true. So don't allow tiger to
        # appear lost until it's tiger's go
        if not mobile_tigers:
            mobile_tigers = int(self.goats_go())

        if self.goats_eaten() >= NUM_EATEN_LOSE:
            score = 100
        elif not mobile_tigers:
            score = -100
        else:
            score = 10 * self.goats_eaten() + mobile_tigers
        if self.goats_go():
            score *= -1
        return score

    @log
    def is_over(self):
        return self.tiger_wins() or self.goat_wins() or self.is_draw()


if __name__ == "__main__":
    kw1, kw2 = {}, {}
    if 1:
        kw1 = {'tt': TT()}
        kw2 = {'tt': TT()}
    goat_ai = Negamax(6, **kw1)
    tiger_ai = Negamax(6, **kw2)
    game = TigerAndGoat([AI_Player(goat_ai, 'goat'), AI_Player(tiger_ai, 'tiger')])

    start = time.monotonic()
    game.play(500)

    duration = time.monotonic() - start
    move_duration = duration / game.nmove
    print('duration', f'{duration:.1f} secs, {move_duration:.2f} secs / move')

#  8: duration 172 seconds
#  9:  99 moves / 360 seconds
# 10: 166 moves / 1011 seconds

File: browser/src/engine/game.ts
// A plain-TypeScript port of server/game.py so the whole game can run in the
// browser with no server. The move rules here mirror the Python engine exactly
// (verified by the parity fixtures in engine.test.ts).
import { JUMPS_GRAPH, STEPS_GRAPH } from '../graph';

export const GOAT_PLAYER = 1;
export const TIGER_PLAYER = 2;
export const NUM_GOATS = 20;
export const NUM_EATEN_LOSE = 5;
export const STARTING_TIGERS = [0, 4, 20, 24];

// Plain adjacency tables derived from the shared Immutable graphs.
const STEPS: number[][] = [];
const JUMPS: number[][] = [];
for (let n = 0; n < 25; n++) {
  STEPS[n] = (STEPS_GRAPH.get(n)?.toArray() ?? []) as number[];
  JUMPS[n] = (JUMPS_GRAPH.get(n)?.toArray() ?? []) as number[];
}

export const STEP_NEIGHBOURS = STEPS;
export const JUMP_NEIGHBOURS = JUMPS;

export type Move = number[]; // length 1 (place), 2 (step), or 3 (jump)
export type Position = [number[], number[]]; // [sortedTigers, sortedGoats]

export interface EngineState {
  tigers: Set<number>;
  goats: Set<number>;
  goatsToPlace: number;
  currentPlayer: number;
  history: Position[]; // canonical positions, most recent last
}

const sorted = (s: Set<number>): number[] => [...s].sort((a, b) => a - b);
const key = (p: Position): string => `${p[0].join(',')}|${p[1].join(',')}`;

export const canonical = (state: EngineState): Position => [
  sorted(state.tigers),
  sorted(state.goats),
];

export function newGame(): EngineState {
  const state: EngineState = {
    tigers: new Set(STARTING_TIGERS),
    goats: new Set(),
    goatsToPlace: NUM_GOATS,
    currentPlayer: GOAT_PLAYER,
    history: [],
  };
  state.history.push(canonical(state));
  return state;
}

export function stateFromHistory(
  playerNum: number,
  goatsToPlace: number,
  history: number[][][],
): EngineState {
  const last = history[history.length - 1];
  return {
    tigers: new Set(last[0]),
    goats: new Set(last[1]),
    goatsToPlace,
    currentPlayer: playerNum,
    history: history.map((p) => [
      [...p[0]].sort((a, b) => a - b),
      [...p[1]].sort((a, b) => a - b),
    ]),
  };
}

const occupied = (state: EngineState): Set<number> =>
  new Set([...state.tigers, ...state.goats]);

function repetitionCount(state: EngineState): number {
  if (state.goatsToPlace > 0) {
    return 1;
  }
  // The latest history entry is the current position with the current side to
  // move; positions alternate side to move, so step back two at a time.
  const current = key(canonical(state));
  let count = 0;
  for (let i = state.history.length - 1; i >= 0; i -= 2) {
    if (key(state.history[i]) === current) {
      count += 1;
    }
  }
  return count;
}

export function isThreefold(state: EngineState): boolean {
  return state.goatsToPlace === 0 && repetitionCount(state) >= 3;
}

export function applyToSets(move: Move, tigers: Set<number>, goats: Set<number>): void {
  if (move.length === 1) {
    goats.add(move[0]);
  } else if (move.length === 2) {
    const [src, dest] = move;
    if (tigers.has(src)) {
      tigers.delete(src);
      tigers.add(dest);
    } else {
      goats.delete(src);
      goats.add(dest);
    }
  } else {
    const [src, eaten, dest] = move;
    tigers.delete(src);
    tigers.add(dest);
    goats.delete(eaten);
  }
}

export function tigerCaptures(state: EngineState): Move[] {
  const occ = occupied(state);
  const moves: Move[] = [];
  for (const t of state.tigers) {
    for (const dest of JUMPS[t]) {
      const eaten = (t + dest) / 2;
      if (state.goats.has(eaten) && !occ.has(dest)) {
        moves.push([t, eaten, dest]);
      }
    }
  }
  return moves;
}

function tigerSteps(state: EngineState): Move[] {
  const occ = occupied(state);
  const moves: Move[] = [];
  for (const t of state.tigers) {
    for (const dest of STEPS[t]) {
      if (!occ.has(dest)) {
        moves.push([t, dest]);
      }
    }
  }
  return moves;
}

function goatPlacements(state: EngineState): Move[] {
  const occ = occupied(state);
  const moves: Move[] = [];
  for (let pos = 0; pos < 25; pos++) {
    if (!occ.has(pos)) {
      moves.push([pos]);
    }
  }
  return moves;
}

function goatSteps(state: EngineState): Move[] {
  const occ = occupied(state);
  const moves: Move[] = [];
  for (const g of state.goats) {
    for (const dest of STEPS[g]) {
      if (!occ.has(dest)) {
        moves.push([g, dest]);
      }
    }
  }
  return moves;
}

export function possibleMoves(state: EngineState): Move[] {
  if (state.currentPlayer === TIGER_PLAYER) {
    return [...tigerCaptures(state), ...tigerSteps(state)];
  }
  return state.goatsToPlace > 0 ? goatPlacements(state) : goatSteps(state);
}

export function mobileTigers(state: EngineState): number {
  const occ = occupied(state);
  const sources = new Set<number>();
  for (const t of state.tigers) {
    let mobile = JUMPS[t].some(
      (dest) => state.goats.has((t + dest) / 2) && !occ.has(dest),
    );
    if (!mobile) {
      mobile = STEPS[t].some((dest) => !occ.has(dest));
    }
    if (mobile) {
      sources.add(t);
    }
  }
  return sources.size;
}

export const goatsEaten = (state: EngineState): number =>
  NUM_GOATS - state.goatsToPlace - state.goats.size;

const tigersGo = (state: EngineState): boolean => state.currentPlayer === TIGER_PLAYER;
const goatsGo = (state: EngineState): boolean => state.currentPlayer === GOAT_PLAYER;

export const tigerWins = (state: EngineState): boolean =>
  goatsEaten(state) >= NUM_EATEN_LOSE;

export const goatWins = (state: EngineState): boolean =>
  tigersGo(state) && mobileTigers(state) === 0;

export function isDraw(state: EngineState): boolean {
  if (state.goatsToPlace > 0) {
    return false;
  }
  if (isThreefold(state)) {
    return true;
  }
  if (tigersGo(state)) {
    return false;
  }
  return possibleMoves(state).length === 0;
}

export const isOver = (state: EngineState): boolean =>
  tigerWins(state) || goatWins(state) || isDraw(state);

export function resultName(state: EngineState): string {
  if (tigerWins(state)) {
    return 'tiger wins';
  }
  if (goatWins(state)) {
    return 'goat wins';
  }
  if (isThreefold(state)) {
    return 'draw (threefold repetition)';
  }
  if (isDraw(state)) {
    return "draw (as goat can't move)";
  }
  return '';
}

export function scoring(state: EngineState): number {
  // Written from the tiger's point of view, then flipped for the goat.
  let mobile = mobileTigers(state);
  if (mobile === 0 && goatsGo(state)) {
    mobile = 1;
  }
  let score: number;
  if (tigerWins(state)) {
    score = 100;
  } else if (goatWins(state)) {
    score = -100;
  } else if (isDraw(state)) {
    score = 0;
  } else {
    score = 10 * goatsEaten(state) + mobile;
  }
  return goatsGo(state) ? -score : score;
}

// makeMove returns the eaten goat (or null) so unmakeMove can restore it.
export function makeMove(state: EngineState, move: Move): number | null {
  let captured: number | null = null;
  applyToSets(move, state.tigers, state.goats);
  if (move.length === 1) {
    state.goatsToPlace -= 1;
  } else if (move.length === 3) {
    captured = move[1];
  }
  state.history.push(canonical(state));
  state.currentPlayer =
    state.currentPlayer === GOAT_PLAYER ? TIGER_PLAYER : GOAT_PLAYER;
  return captured;
}

export function unmakeMove(
  state: EngineState,
  move: Move,
  captured: number | null,
): void {
  state.currentPlayer =
    state.currentPlayer === GOAT_PLAYER ? TIGER_PLAYER : GOAT_PLAYER;
  state.history.pop();
  if (move.length === 1) {
    state.goats.delete(move[0]);
    state.goatsToPlace += 1;
  } else if (move.length === 2) {
    const [src, dest] = move;
    if (state.tigers.has(dest)) {
      state.tigers.delete(dest);
      state.tigers.add(src);
    } else {
      state.goats.delete(dest);
      state.goats.add(src);
    }
  } else {
    const [src, , dest] = move;
    state.tigers.delete(dest);
    state.tigers.add(src);
    if (captured !== null) {
      state.goats.add(captured);
    }
  }
}
